recirq/quantum_chess/quantum_board.py

```diff
@@ -325,36 +337,113 @@ def sample(self, num_samples: int) -> List[int]:
             rtn = rtn + samples
         return rtn[:num_samples]
 
-    def _generate_accumulations(self, repetitions: int = 1000) -> None:
+    def _set_full_empty_squares_from_probability(self) -> None:
+        self.full_squares = 0
+        self.empty_squares = 0
+        for i, p in enumerate(self.probabilities):
+            if p == 1:
+                self.full_squares = set_nth_bit(i, self.full_squares, True)
+            if p == 0:
+                self.empty_squares = set_nth_bit(i, self.empty_squares, True)
+
+    def _generate_accumulations(self, repetitions: int = 1000, use_cache: bool = False) -> None:
         """ Samples the state and generates the accumulated 
         probabilities of each square, empty_squares, and full_squares.
         """
-        self.probabilities = [0] * 64
-        self.full_squares = (1 << 64) - 1
-        self.empty_squares = (1 << 64) - 1
 
+        if use_cache and self.move_history:
+            last_move = self.move_history[-1]
+            if self.move_history_probabilities_cache[-1][0] >= repetitions:
+                self.probabilities = self.move_history_probabilities_cache[-1][1].copy()
+                self._set_full_empty_squares_from_probability()
+                return
+            previous_move_in_cache = len(self.move_history) > 1 and self.move_history_probabilities_cache[-2][0] >= repetitions
+            is_first_move = len(self.move_history) == 1
+            cache_key = cache_key_from_move(last_move, repetitions)
+            if (previous_move_in_cache or is_first_move) and self._caching_supported(last_move) and cache_key in self.cache:
+                if previous_move_in_cache:
+                    previous_probability = self.move_history_probabilities_cache[len(self.move_history) - 2][1]
+                else:
+                    previous_probability = [0] * 64
+                    for i in range(64):
+                        # Assume initial state is classical
+                        previous_probability[i] = nth_bit_of(i, self.state)
+
+                probs = self._apply_cache(previous_probability, last_move,
+                    self.cache[cache_key])
+                self.probabilities = probs
+                self._set_full_empty_squares_from_probability()
+                self.move_history_probabilities_cache.append((repetitions, probs.copy()))
+                del self.cache[cache_key]
```

Comment:
Comment on why the cache entry is deleted?

Reply:
Added a comment.